gobblin-modules/google-ingestion/src/main/java/gobblin/ingestion/google/webmaster/GoogleWebmasterExtractor.java

```diff
@@ -0,0 +1,221 @@
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Queue;
+import org.joda.time.DateTime;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static gobblin.configuration.ConfigurationKeys.*;
+import static gobblin.source.extractor.extract.google.GoogleCommonKeys.*;
+
+
+public class GoogleWebmasterExtractor implements Extractor<String, String[]> {
+  private final static Logger LOG = LoggerFactory.getLogger(GoogleWebmasterExtractor.class);
+  private final static Splitter splitter = Splitter.on(",").omitEmptyStrings().trimResults();
+  private final String _schema;
+  private final WorkUnitState _wuState;
+  private final int _size;
+  private final static DateTimeFormatter dateFormatter = DateTimeFormat.forPattern("yyyy-MM-dd");
+  private final static DateTimeFormatter watermarkFormatter = DateTimeFormat.forPattern("yyyyMMddHHmmss");
+  private Queue<GoogleWebmasterExtractorIterator> _iterators = new ArrayDeque<>();
+  /**
+   * Each element keeps a mapping from API response order to output schema order.
+   * The array index matches the order of API response.
+   * The array values matches the order of output schema.
+   */
+  private Queue<int[]> _positionMaps = new ArrayDeque<>();
+
+  private final DateTime _startDate;
+  private final DateTime _endDate;
+  private boolean _successful = false;
+
+  public GoogleWebmasterExtractor(WorkUnitState wuState, long lowWatermark, long highWatermark,
+      Map<String, Integer> columnPositionMap, List<GoogleWebmasterFilter.Dimension> requestedDimensions,
+      List<GoogleWebmasterDataFetcher.Metric> requestedMetrics) throws IOException {
+    this(wuState, lowWatermark, highWatermark, columnPositionMap, requestedDimensions, requestedMetrics,
+        new GoogleWebmasterDataFetcherImpl(wuState.getProp(GoogleWebMasterSource.KEY_PROPERTY), getCredential(wuState),
+            wuState.getProp(ConfigurationKeys.SOURCE_ENTITY), getHotStartJobs(wuState)));
+  }
+
+  private static List<ProducerJob> getHotStartJobs(WorkUnitState wuState) {
+    String hotStartString = wuState.getProp(GoogleWebMasterSource.KEY_REQUEST_HOT_START, "");
+    if (!hotStartString.isEmpty()) {
+      return SimpleProducerJob.deserialize(hotStartString);
+    }
+    return new ArrayList<>();
+  }
+
+  private static Credential getCredential(WorkUnitState wuState) {
+    String scope = wuState.getProp(GoogleCommonKeys.API_SCOPES, WebmastersScopes.WEBMASTERS_READONLY);
+    Preconditions.checkArgument(
+        Objects.equals(WebmastersScopes.WEBMASTERS_READONLY, scope) || Objects.equals(WebmastersScopes.WEBMASTERS,
+            scope), "The scope for WebMaster must either be WEBMASTERS_READONLY or WEBMASTERS");
+
+    String credentialFile = wuState.getProp(SOURCE_CONN_PRIVATE_KEY);
```

Comment:
The SOURCE_CONN_PRIVATE_KEY here is a path to the private key file, which will be kept at a place with limited access. So does the master password file in your case. But thank you for sharing this because my next project will need this feature.